summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -51,24 +49,25 @@ class AggregatorOutputCollector[K, V: Semigroup](
     The method is invoked to handle the flushed cache caused by
     exceeding the memoryLimit, which is called within add method.
    */
-  private def emitData[K, V](cache: Future[Traversable[(Int, CMap[K, V])]], streamId: String): List[Int] = {
+  private def emitData[K, V](cache: Future[TraversableOnce[(Int, CMap[K, V])]], streamId: String): JList[Integer] = {
     val flushedTups = Await.result(cache)
     val messageIdsTracker = streamMessageIdTracker(streamId)
-    val returns = flushedTups.toList
-      .map {
-        case (k, v) =>
-          val messageIds = messageIdsTracker.remove(k)
-          val list = List(k, v).asJava.asInstanceOf[JList[AnyRef]]
-          callEmit(messageIds, list, streamId)
-      }
-    returns.flatMap { x => { if (x != null) x.asScala.toList.asInstanceOf[List[Int]] else List[Int]() } }
+    val returns = flushedTups.map {
+      case (k, v) =>
+        val messageIds = messageIdsTracker.remove(k)
+        val list = List(k, v).asJava.asInstanceOf[JList[AnyRef]]
+        callEmit(messageIds, list, streamId)
+    }
+    val result = new java.util.ArrayList[Integer]()
+    returns.filter(_ != null).foreach { result.addAll(_) }
+    result
   }
 
   /*
    This is a wrapper method to call the emit with appropriate signature
    based on the arguments.
   */
-  private def callEmit(messageIds: Option[Any], list: JList[AnyRef], stream: String): JList[Integer] = {
+  private def callEmit(messageIds: Option[TraversableOnce[AnyRef]], list: JList[AnyRef], stream: String): JList[Integer] = {
```

Comment:
.emit sends an object as the messageId. It is acked back by passing the same object, which should be handled on the spout.